### PR TITLE
Getting ALL from API

### PR DESCRIPTION
**Pull request by MaximBadanov**
- переделал все на экране с выбором жанра, теперь все жанры мы получаем из из сети
- разобрался с вьюмоделями, когда один экран включает другой
- разобрался с кнопкой и передачей данных в ResultView
- чуть поменял шрифт названия фильма, тк (см. скрины)
Simulator Screenshot - iPhone 15 Pro - 2024-06-07 at 00 49 27

### CODE REVIEW

File: swiftui_project/swiftui_project/Supports&Extensions/StringExtension/StringExtension.swift [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            return .green
+        case "horror":
+            return .teal
+        case "documentary":
+            return .indigo
+        case "drama":
+            return .blue
+        case "family":
+            return .pink
+        case "history":
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/Supports&Extensions/StringExtension/StringExtension.swift:73]: For few cases you can use
```swift
case “first”, “second”:
    ….
```

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/ViewModel/GenreScreenViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import Combine
+import SwiftUI
+import Foundation
+
+class GenreScreenViewModel: ObservableObject {
+    @Published var movie: MovieResponseModel?
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/ViewModel/GenreScreenViewModel.swift:6]: private(set)

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/ViewModel/GenreScreenViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    @Published var genresId: [String] = []
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/ViewModel/GenreScreenViewModel.swift:7]: The same

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                            modelToParse: MovieResponse.self
+                        ),
+                        genreIDs: viewModel.genresId
+                    )
+                }) {
+                    Text("Get Movie")
+                }
+                .buttonStyle(.primaryStyle)
+                NavigationLink("Continue",
+                               destination: ResultView(title: viewModel.movie?.title ?? "no title", poster: viewModel.movie?.poster ?? "no poster"))
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift:37]: Better on this side send optional value

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                )
+            GenresViewWithScroll()
+            Spacer(minLength: UISize.size8)
+            HStack(spacing: UISize.size8) {
+                Spacer(minLength: UISize.size8)
+                    .padding(.horizontal, UISize.size24)
+                //                от сюда
+                Button(action: {
+                    viewModel.fetchMoviesByGenre(
+                        requestModel: RequestModel(
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift:24]: RequestModel MUST NOT BE on view side

File: swiftui_project/swiftui_project/ScreenWithGenres/CarouselOfGenres/ViewModel/GenresViewWithScrollViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+                case .failure(let error):
+                    print("Finished with error: \(error)")
+                }
+            }, receiveValue: { data in
+                DispatchQueue.main.async {
+                    self.fetchedGenres = data.genres
+                }
+            })
+    }
+
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/CarouselOfGenres/ViewModel/GenresViewWithScrollViewModel.swift:43]: Extra

File: swiftui_project/swiftui_project/NetworkService/NetworkService.swift [hunk 7]
```diff
@@ -17,4 +16,33 @@ class NetworkService: NetworkServiceProtocol {
             .decode(type: T.self, decoder: JSONDecoder())
             .eraseToAnyPublisher()
     }
+
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/NetworkService/NetworkService.swift:19]: Why request func only for movies?
You need do generic func for any request, such as movie fetching or weather from another API

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift (hunk 4 above)
Comment (MaximBadanov) [reply swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift:37]: I  don't understand, can you explain this a little bit more

File: swiftui_project/swiftui_project/NetworkService/NetworkService.swift (hunk 7 above)
Comment (MaximBadanov) [reply swiftui_project/swiftui_project/NetworkService/NetworkService.swift:19]: I will refactor this after I finish with throwing data between 2 view models(

File: swiftui_project/swiftui_project/Supports&Extensions/StringExtension/StringExtension.swift (hunk 1 above)
Comment (MaximBadanov) [reply swiftui_project/swiftui_project/Supports&Extensions/StringExtension/StringExtension.swift:73]: but I have to many cases 🤷🏻‍♂️

File: swiftui_project/swiftui_project/ResultView/MovieView.swift [hunk 8]
```diff
@@ -3,8 +3,9 @@ import SwiftUI
 struct MovieView: View {
     let title: String
     let poster: String
+    let genres: String
     
-    private let aspectRatioPoster: Double = 2.4/3.5
+    private let aspectRatioPoster: Double = 2.4/3.2
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ResultView/MovieView.swift:8]: You dont need it, just ude maxWidth for frame and fit aspect ratio without value

File: swiftui_project/swiftui_project/ScreenWithGenres/CarouselOfGenres/ViewModel/GenresViewWithScrollViewModel.swift [hunk 9]
```diff
@@ -52,7 +52,6 @@ extension GenresViewWithScrollViewModel: GenresViewWithScrollViewModelProtocol {
             selectedGenres.append(id)
         }
         print("Коды выбранных жанров: \(selectedGenres)")
+        UserDefaults.standard.setValue(selectedGenres, forKey: "selectedGenres")
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/CarouselOfGenres/ViewModel/GenresViewWithScrollViewModel.swift:55]: We don’t need to save in memory the genres, it’s not a cross launching value

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift [hunk 10]
```diff
@@ -21,20 +21,19 @@ struct GenresScreen: View {
...
                 }) {
                     Text("Get Movie")
                 }
                 .buttonStyle(.primaryStyle)
                 NavigationLink("Continue",
-                               destination: ResultView(title: viewModel.movie?.title ?? "no title", poster: viewModel.movie?.poster ?? "no poster"))
+                               destination: ResultView(
+                                title: viewModel.movie?.title ?? "no title",
+                                poster: viewModel.movie?.poster ?? "no poster",
+                                genres: viewModel.convertIdsInString(genres:viewModel.movie?.genres ?? [0])
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift:34]: Why are you using properties VM to pass to the VM method again? U can do it in VM side

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/ViewModel/GenreScreenViewModel.swift [hunk 11]
```diff
@@ -3,22 +3,35 @@ import SwiftUI
...
     
+    var genresId: [String]  {
+        UserDefaults.standard.object(forKey: "selectedGenres")  as? [String] ?? [String]()
+    }
     
     init() {
         dataManager = DataManager()
     }
+    
+    func convertIdsInString(genres: [Int]) -> String {
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/ViewModel/GenreScreenViewModel.swift:19]: Better do it on Request side where are you creating Request of Movies

**Review by iosnickse — CHANGES_REQUESTED** [review]
Small changes

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/ViewModel/GenreScreenViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+   private func convertIdsInString(genres: [Int]) -> String {
+        let genreStrings = genres.map { String($0).convertIdToString }
+        return genreStrings.joined(separator: " ")
+    }
+}
+
+extension GenreScreenViewModel: GenreScreenViewModelProtocol {
+    func fetchMoviesByGenre(completion: @escaping (Bool) -> Void) {
```
Comment (iosnickse) [inline swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/ViewModel/GenreScreenViewModel.swift:24]: no need it

File: swiftui_project/swiftui_project/Supports&Extensions/StringExtension/StringExtension.swift (hunk 1 above)
Comment (iosnickse) [reply swiftui_project/swiftui_project/Supports&Extensions/StringExtension/StringExtension.swift:73]: Try to group by colour for few cases

**Review by iosnickse — CHANGES_REQUESTED** [review]

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+struct GenresScreen: View {
+    @ObservedObject private var genreScreenViewModel = GenreScreenViewModel()
+    @ObservedObject private var genresWithScrollViewModel = GenresWithScrollViewModel()
+    @State private var navigateToResult = false
+    
+    var body: some View {
+        NavigationStack {
+            VStack(alignment: .leading, spacing: UISize.size8) {
+                Spacer(minLength: UISize.size8)
```
[the remark this answers is not in the document]
Comment (MaximBadanov) [reply swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift:11]: without it all ui  floats(

File: swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+                Text("Choose Genres")
+                    .padding(.leading, UISize.size24)
+                    .textStyle(size: UISize.size32, weight: .bold)
+                GenresViewWithScroll(viewModel: genresWithScrollViewModel,
+                                     selectedGenres: $genreScreenViewModel.genresId)
+                Spacer(minLength: UISize.size8)
+                HStack(spacing: UISize.size8) {
+                    Spacer(minLength: UISize.size8)
```
[the remark this answers is not in the document]
Comment (MaximBadanov) [reply swiftui_project/swiftui_project/ScreenWithGenres/GenresScreen/GenresScreen.swift:19]: same thing

**Pull request merged**